### PR TITLE
[autograd] Do not detach when unpacking tensors that do not require grad 

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127959
* #129262
* #128545
* #125795

In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops.

### CODE REVIEW

File: test/test_nestedtensor.py [hunk 1]
```diff
@@ -4496,6 +4501,31 @@ def f(values, offsets):
...
+        checkpoint(fn, values, offsets, use_reentrant=False).backward()
+
+        context_fn = partial(
+            create_selective_checkpoint_contexts,
+            [
+                torch.ops.aten.cumsum.default,
+            ]
+        )
+
+        def fn(values, offsets):
```
Comment (YuqingJ) [inline test/test_nestedtensor.py:4522]: Do you mean def fn(values, lengths) ?

Comment (soulitzer) [reply test/test_nestedtensor.py:4522]: Yup, good catch

File: test/test_nestedtensor.py [hunk 2]
```diff
@@ -4496,6 +4501,31 @@ def f(values, offsets):
...
+                torch.ops.aten.cumsum.default,
+            ]
+        )
+
+        def fn(values, lengths):
+            offsets = F.pad(lengths, pad=(1, 0)).cumsum(dim=0)
+            nt = convert_jagged_to_nested_tensor(values, offsets, max_length=4)
+            return convert_nt_to_jagged(nt).sum()
+
+        checkpoint(fn, values, lengths, use_reentrant=False, context_fn=context_fn).backward()
```
Comment (YuqingJ) [inline test/test_nestedtensor.py:4527]: nit: maybe add another check to make sure values.grad is not None?

**Review by YuqingJ — APPROVED** [review]

**Comment by YuqingJ** [comment]
Ready to merge again?

**Pull request closed without merging**